driver/src/price_finding/naive_solver.rs

```diff
@@ -218,33 +209,95 @@ fn order_with_buffer_for_fee(order: &Order, fee: &Option<Fee>) -> Order {
     }
 }
 
+/// Normalizes a price base on the pre-normalized fee price.
+fn normalize_price(price: u128, pre_normalized_fee_price: u128) -> Option<u128> {
+    // upcast to u256 to avoid overflows
+    checked_u256_to_u128(
+        (u128_to_u256(price) * u128_to_u256(BASE_PRICE))
+            .ceiled_div(u128_to_u256(pre_normalized_fee_price)),
+    )
+}
+
+/// Calculate the executed sell amount from the fee, executed buy amount, and
+/// the buy and sell prices of the traded tokens.
+fn executed_sell_amount(fee: &Fee, exec_buy_amt: u128, buy_price: u128, sell_price: u128) -> u128 {
+    let fee_denominator = (1.0 / fee.ratio) as u128;
+    u256_to_u128(
+        (((u128_to_u256(exec_buy_amt) * u128_to_u256(buy_price))
+            / u128_to_u256(fee_denominator - 1))
+            * u128_to_u256(fee_denominator))
+            / u128_to_u256(sell_price),
+    )
+}
+
+/// Calculate the executed buy amount from the fee, executed sell amount, and
+/// the buy and sell prices of the traded tokens. This function returns a `None`
+/// if no value can be found such that `executed_sell_amount(result, pb, ps) ==
+/// xs`. This function acts as an inverse to `executed_sell_amount`.
+fn executed_buy_amount(
+    fee: &Fee,
+    exec_sell_amt: u128,
+    buy_price: u128,
+    sell_price: u128,
+) -> Option<u128> {
+    let fee_denominator = (1.0 / fee.ratio) as u128;
+    let exec_buy_amt = u256_to_u128(
+        (((u128_to_u256(exec_sell_amt) * u128_to_u256(sell_price))
+            / u128_to_u256(fee_denominator))
+            * u128_to_u256(fee_denominator - 1))
+            / u128_to_u256(buy_price),
+    );
+
+    // we need to account for rounding errors here, since this function is
+    // essentially an inverse of `executed_sell_amount`; so find the maximum
+    // error that `v` can have and find a value that works
+    // TODO(nlordell): verify the maths
```

Comment:
```suggestion
    // TODO(nlordell): verify the math
```

Reply:
It was meant to be sarcastic. I should probably remove that comment. I was hoping one of you guys could take a look at the logic there and make sure it makes sense.

Reply:
Will do

Reply:
Removed the TODO :smile: